src/utils/rfq_blocked_address_utils.ts

```diff
@@ -0,0 +1,64 @@
+import { Connection } from 'typeorm/connection/Connection';
+
+import { BlockedAddressEntity } from '../entities/BlockedAddressEntity';
+
+/**
+ * RfqBlockedAddressUtils helps manage the RFQ blocked addresses
+ */
+export class RfqBlockedAddressUtils {
+    private _blocked: Set<string>;
+    private _expiresAt: number;
+    private _updatePromise: Promise<void> | undefined;
+    private _updating: boolean;
+    private readonly _ttlMs: number;
+
+    constructor(private readonly _connection: Connection, initialBlacklist: Set<string>, ttlMs: number) {
+        this._blocked = initialBlacklist;
+        this._ttlMs = ttlMs;
+        this._updating = false;
+        this._expiresAt = Date.now().valueOf(); // cache expires immediately
+    }
+
+    /**
+     * isBlocked returns whether an address is blocked from the cache
+     * NOTE: In the background, it also updates the blocked set if the cache is expired
+     */
+    public isBlocked(address: string): boolean {
+        if (Date.now().valueOf() > this._expiresAt && !this._updating) {
+            // If expired, update in the background
+            this._updatePromise = this._updateBlockedSetAsync();
+        }
+
+        // Return cached value, even if stale
+        return this._blocked.has(address.toLowerCase());
+    }
+
+    /**
+     * isBlockedAsync returns whether an address is blocked. The value is as "fresh" as the TTL allows
+     */
+    public async isBlockedAsync(address: string): Promise<boolean> {
+        // If already updating, await until done
+        if (this._updating) {
+            await this._updatePromise;
+        }
+
+        // Check if the TTL has been exceeded
+        if (Date.now().valueOf() > this._expiresAt && !this._updating) {
+            await this._updateBlockedSetAsync();
+        }
+
+        // Guaranteed to be a fresh value
+        return this._blocked.has(address.toLowerCase());
+    }
+
+    /**
+     * Updates the blocked set of addresses
+     */
+    private async _updateBlockedSetAsync(): Promise<void> {
+        this._updating = true;
+        const blockedAddresses = await this._connection.getRepository(BlockedAddressEntity).find();
```

Comment:
can we add a `LIMIT` here just to be safe? just to avoid our system from blowing up and make it more determinisric

Reply:
For example:
- Limit query to 5000 addresses
- If query returns 5000 addresses, log an error (so that we are informed that the blocked entities have passed a certain threshold)

Reply:
Do we read from our slave node in this case? or do we use the master node? do we know?

Reply:
✅  limit 5000 with logging

regarding the replica, it depends on if the master/slave configuration is enabled: https://typeorm.io/#/multiple-connections/replication

In our case, we do use the master/slave configuration so _all_ `find` queries go to one of the read replicas by default, and all inserts and updates will go to the write master.